### ISSUE TITLE
Fold `setjmp` into Cranelift-generated entry trampolines to WebAssembly

### ISSUE THREAD

**Issue by alexcrichton** [bytecodealliance/wasmtime#3927]
This issues comes out of a discussion that @lukewagner, @fitzgen, and I were having recently. We were thinking again about how Wasmtime implements calls into WebAssembly and about some of the overhead associated with that. Currently it's suprisingly expensive relative to wasm->host transitions, where host->wasm is on the order of 20-30ns where wasm->host is on the order of 3-5ns.

One of the major costs of entering WebAssembly is that we have to call `setjmp`. Not only is [`setjmp` complicated since it's platform-specific](https://github.com/bytecodealliance/wasmtime/blob/a68f44a924c4dac120d12cf59f08a6c0863c6837/crates/runtime/src/helpers.c#L5-L43) but as seen there it's also written in C. We can't call setjmp from Rust (since it "returns twice" and the Rust compiler doesn't inform LLVM of that, meaning optimizations could go awry) which means entering WebAssembly is even further de-optimized because all arguments must pass through the stack. [This closure](https://github.com/bytecodealliance/wasmtime/blob/a68f44a924c4dac120d12cf59f08a6c0863c6837/crates/runtime/src/traphandlers.rs#L184) captures all arguments into WebAssembly and is forced to be on the stack as we pass a single pointer to C which is [called back](https://github.com/bytecodealliance/wasmtime/blob/a68f44a924c4dac120d12cf59f08a6c0863c6837/crates/runtime/src/traphandlers.rs#L189-L194).

Another further complication with this current strategy of entering WebAssembly is that in a future world with the wasm exceptions proposal  whatever is chosen to implement exceptions at the cranelift level is highly unlikely to be exposed in the full fidelity required to native stable Rust, meaning that we couldn't actually write a "catch" block in Rust (and probably not C). 

To solve all these issues, @lukewagner mentioned we could do something like SpiderMonkey which is to have specialized entry trampolines into WebAssembly code. Currently our trampolines are primarily just converting from a dynamic stack-based layout to a particular System-V ABI signature, which isn't really all that interesting. Instead, though, we could specifically have a trampoline that receives the appropriate arguments, sets up a "catch" frame, and then enters the desired WebAssembly code. This could have a number of benefits:

* Nothing is forced to be on the stack since we never cross between Rust & C. Instead Rust, when using `TypedFunc`, would make a System-V call to this entry trampoline and the entry trampoline would indicate via its return value whether a trap was caught. Note, though, that we may still want to store params/results on the stack for other reasons, such as communicating results since this trampoline would always have at least one result (whether the function trapped or not).
* We get to entirely define `setjmp`  and/or the trap exception protocol within Cranelift. This entry trampoline would implement `setjmp`, or at least the pieces necessary, or whatever exception handling implementation we get around to using in Cranelift.
* This could further reduce the need for external C helpers which moves Wasmtime a bit closer to having 0 C dependencies.

Another possible idea is that currently trampolines are one-per-function-signature which means that they always contain an indirect call to a target. Instead we could also explore a scheme where we have one-per-export which would enable the trampoline to statically call into the correct export (no indirect function call necessary) which is another route to possibly optimize this.

---

The implementation of setjmp/longjmp in Cranelift is likely to be pretty nontrivial for this which is why I wanted to open an issue on this and let it get some feedback before implementing. I also don't think that this is super pressing at this time to the point that we should implement it, but it's good to have in our back pocket if we run into issues with the overhead of host->wasm transitions. I'm not actually sure how we'd implement setjmp/longjmp in Cranelift (e.g. expose it and represent it in clif) myself. Implementation-wise we'd probably want to at least take inspiration if not scrutinize the SpiderMonkey implementation since we don't need a general setjmp/longjmp mechanism, only one that works for wasm traps.

**Comment by lukewagner** [comment]
Agreed that you can get away with something much simpler than a general-purpose setjmp/longjmp implementation.

Just putting out some info and links for how SpiderMonkey does this in case anyone is interested later:
* When it's time to throw, the "normal" compiled wasm code jumps to the throw stub generated by [`GenerateThrowStub`](https://searchfox.org/mozilla-central/source/js/src/wasm/WasmStubs.cpp#2866)
* The throw stub calls the C++ function [`WasmHandleThrow`](https://searchfox.org/mozilla-central/source/js/src/wasm/WasmBuiltins.cpp#500) which uses the stack iterator to determine the fp/sp to unwind to (by walking the stack frame structure maintained by the entry trampoline and normal calls)
* The throw stub executes a careful, but ultimately not too complicated, [sequence of instructions](https://searchfox.org/mozilla-central/source/js/src/wasm/WasmStubs.cpp#2909-2959) to update to the sp found by the previous step (which points to the return address pushed by the original call from the entry trampoline into normal wasm code) and then executes a `ret`.
* Instead of using the ABI's normal return-value register to communicate the error (which are already used for the non-exceptional return value), the convention SpiderMonkey uses is storing `0xbad` in the fp register (right before returning from the throw stub to the entry trampoline), which the entry trampoline [immediately branches on](https://searchfox.org/mozilla-central/source/js/src/wasm/WasmStubs.cpp#1294-1298).

Ultimately, because the sp of the entry trampoline is restored to the same offset expected upon normal (non-exceptional) return, the entry trampoline doesn't need to do much specially: it simply places the data it needs for returning (exceptionally and normally) in the stack frame and branches right after the call into normal wasm code to detect the exceptional return.
